frida_mode/src/complog/complog_x64.c

```diff
@@ -0,0 +1,367 @@
+
+    case X86_REG_R9D:
+      return ctx->r9 & GUM_INT32_MASK;
+
+    case X86_REG_R10D:
+      return ctx->r10 & GUM_INT32_MASK;
+
+    case X86_REG_R11D:
+      return ctx->r11 & GUM_INT32_MASK;
+
+    case X86_REG_R12D:
+      return ctx->r12 & GUM_INT32_MASK;
+
+    case X86_REG_R13D:
+      return ctx->r13 & GUM_INT32_MASK;
+
+    case X86_REG_R14D:
+      return ctx->r14 & GUM_INT32_MASK;
+
+    case X86_REG_R15D:
+      return ctx->r15 & GUM_INT32_MASK;
+
+    case X86_REG_EIP:
+      return ctx->rip & GUM_INT32_MASK;
+
+    case X86_REG_RAX:
+      return ctx->rax;
+
+    case X86_REG_RCX:
+      return ctx->rcx;
+
+    case X86_REG_RDX:
+      return ctx->rdx;
+
+    case X86_REG_RBX:
+      return ctx->rbx;
+
+    case X86_REG_RSP:
+      return ctx->rsp;
+
+    case X86_REG_RBP:
+      return ctx->rbp;
+
+    case X86_REG_RSI:
+      return ctx->rsi;
+
+    case X86_REG_RDI:
+      return ctx->rdi;
+
+    case X86_REG_R8:
+      return ctx->r8;
+
+    case X86_REG_R9:
+      return ctx->r9;
+
+    case X86_REG_R10:
+      return ctx->r10;
+
+    case X86_REG_R11:
+      return ctx->r11;
+
+    case X86_REG_R12:
+      return ctx->r12;
+
+    case X86_REG_R13:
+      return ctx->r13;
+
+    case X86_REG_R14:
+      return ctx->r14;
+
+    case X86_REG_R15:
+      return ctx->r15;
+
+    case X86_REG_RIP:
+      return ctx->rip;
+
+    default:
+      FATAL("Failed to read register: %d", reg);
+      return 0;
+
+  }
+
+}
+
+static guint64 complog_read_mem(GumX64CpuContext *ctx, x86_op_mem *mem) {
+
+  guint64 base = 0;
+  guint64 index = 0;
+  guint64 address;
+
+  if (mem->base != X86_REG_INVALID) base = complog_read_reg(ctx, mem->base);
+
+  if (mem->index != X86_REG_INVALID) index = complog_read_reg(ctx, mem->index);
+
+  address = base + (index * mem->scale) + mem->disp;
+  return address;
+
+}
+
+static void complog_handle_call(GumCpuContext *context, guint64 target) {
+
+  guint64 address = complog_read_reg(context, X86_REG_RIP);
+  guint64 rdi = complog_read_reg(context, X86_REG_RDI);
+  guint64 rsi = complog_read_reg(context, X86_REG_RSI);
+
+  void *ptr1 = GSIZE_TO_POINTER(rdi);
+  void *ptr2 = GSIZE_TO_POINTER(rsi);
+
+  if (!complog_is_readable(ptr1, 32) || !complog_is_readable(ptr2, 32)) return;
+
+  uintptr_t k = address;
+
+  k = (k >> 4) ^ (k << 8);
+  k &= CMP_MAP_W - 1;
+
+  __afl_cmp_map->headers[k].type = CMP_TYPE_RTN;
+
+  u32 hits = __afl_cmp_map->headers[k].hits;
+  __afl_cmp_map->headers[k].hits = hits + 1;
+
+  __afl_cmp_map->headers[k].shape = 31;
+
+  hits &= CMP_MAP_RTN_H - 1;
+  gum_memcpy(((struct cmpfn_operands *)__afl_cmp_map->log[k])[hits].v0, ptr1,
+             32);
+  gum_memcpy(((struct cmpfn_operands *)__afl_cmp_map->log[k])[hits].v1, ptr2,
+             32);
+
+}
+
+static guint64 cmplog_get_operand_value(GumCpuContext *context,
+                                        complog_ctx_t *ctx) {
+
+  switch (ctx->type) {
+
+    case X86_OP_REG:
+      return complog_read_reg(context, ctx->reg);
```

Comment:
What about non-64 byte compares?

Reply:
The function `complog_read_reg` deals with comparisons of different sizes. Although a 64bit type is used to represent the value, it is masked and shifted as necessary.